Add a glossary of Teleport terms (#16762)
gravitational/teleport: docs/pages/core-concepts.mdx

---
title: Teleport Core Concepts
description: Learn the the key architectural components that make up Teleport.
---

Here are the core concepts that make up a Teleport deployment. You will see
these terms often when setting up and managing Teleport, so you should get
familiar with them before following other pages in the documentation.

> ## Authentication connector

Reviewer: I think it would improve readability and comprehension if we pulled out the primary terminologies into its own section at the top and then have an alphabetical list of terminologies in a secondary section. For example, we should pull Teleport Cluster, Teleport Process, and Teleport Service out and put them at the top so users understand the fundamental pieces of Teleport before reading further.

authenticate to Teleport via a Single Sign-On (SSO) solution. 

See our guide to [Authentication Options](./reference/authentication.mdx).

## Certificate authority

Teleport allows services and clients to achieve secure access using SSH and TLS
certificates. 

The components of a **Teleport Cluster**, such as clients and **Teleport
Services**, use these certificates to verify that they are connecting with a
genuine Teleport Service or user. To do so, they check that the certificate was
signed by a **certificate authority**—an entity that issues certificates—that
belongs to your Teleport Cluster. 

The **Teleport Auth Service** includes multiple certificate authorities, which
it uses for different purposes.

Read about how Teleport's certificate authorities work in our [Auth Service
architecture guide](./architecture/authentication.mdx).

## Configuration resource

A **configuration resource** is a document stored on the **Auth Service**
backend that specifies settings for your **Teleport cluster**. Examples include
**roles**, **local users**, and **authentication connectors**

Read more in our [Resource Reference](./reference/resources.mdx).

> ## Machine ID

Reviewer: It's worth mentioning here that Machine ID is commonly used for machine-to-machine and service-to-service use cases.

Teleport Services and the Auth Service in a private network, with only the Proxy
Service's HTTPS port exposed to the internet.

You can run multiple Auth Service instances in a cluster for high availability.

Read our guides to how [authorization](./architecture/authorization.mdx) and
[authentication](./architecture/authentication.mdx) work in Teleport.

## Teleport Cloud

**Teleport Cloud** is a managed deployment of the **Teleport Control Plane**.
Our team at Teleport handles all tasks related to running the **Auth Service**
and **Proxy Service**, including upgrades and certificate management. Each
customer account, known as a **Teleport Cloud tenant**, has its own subdomain of
`.teleport.sh`, e.g., `mytenant.teleport.sh`.

Read more in our [Teleport Cloud
guide](./deploy-a-cluster/teleport-cloud/introduction.mdx).

## Teleport Cluster

The key concept of Teleport's architecture is the **cluster**. A Teleport
cluster consists of the **Teleport Auth Service** and **Teleport Proxy Service**,
plus several **Teleport Services** that manage traffic to resources within your
infrastructure, such as Kubernetes clusters and Windows desktops. A Teleport

> cluster also includes **Machine ID** and **Teleport Discover**.

Reviewer: > A Teleport cluster also includes **Machine ID** and **Teleport Discover**

This part didn't make sense to me. I would clarify that Machine ID and Teleport Discover are both services you can run in a cluster.

Reviewer 2: Teleport Discover isn't really a service that you run in the cluster though - is it?

I'm not sure we should even mention discover here. That feels more like an internal name to me. It's not used anywhere in the UI (or docs, to my knowledge)

Author: [This PR](https://github.com/gravitational/teleport/pull/16279) introduces the "Teleport Discovery Service", which maybe we could mention instead of Teleport Discover?

A minimal Teleport Cluster consists of the Teleport Auth Service and Teleport
Proxy Service. In a demo environment, you can run these two services from a
single **Teleport Process**.

## Teleport Control Plane

The Control Plane consists of the **Auth Service** and **Proxy Service**, which
govern cluster-wide access. Other **Teleport Services**, which ensure access to

> specific resources, are not considered part of the Teleport Control Plane.

Reviewer: Let's deprecate the "Teleport Control Plane" terminology altogether in favor or "Teleport Cluster."

## Teleport Enterprise

**Teleport Enterprise** is a paid plan that includes additional features for the
**Teleport Control Plane** and **Teleport Services**, plus a support agreement
with Teleport.

Read more in our [Teleport Enterprise

> guide](./deploy-a-cluster/teleport-enterprise/introduction.mdx).

Reviewer: I'm not a fan of the terminology "Teleport Enterprise" because it implies that Teleport Cloud is not enterprise. This is how I think about our different offerings:

- Teleport Community Edition (this is open source, free, and self-hosted only)
- Teleport Enterprise Edition (this is paid and can be self-hosted or SaaS)
    - Teleport Enterprise Self-Hosted
    - Teleport Enterprise Cloud

>

![Proxy service](../img/architecture/proxy.png)

</Figure>

You can also configure clients to bypass Proxy Service instances and connect to
resources with Teleport-issued certificates directly. 

Read our guide to [how the Teleport Proxy Service
works](./architecture/proxy.mdx).

## Teleport Service

A **Teleport Service** manages access to resources in your infrastructure, such
as Kubernetes clusters, Windows desktops, internal web applications, and
databases. All Teleport Services can run from a single **Teleport Process**, as
can the **Auth Service** and **Proxy Service**:

|Service|Traffic it Proxies|
|---|---|
|[Teleport Database Service](./database-access/introduction.mdx)|TCP traffic in the native protocols of popular databases, including PostgreSQL and MySQL.|
|[Teleport Application Service](./application-access/introduction.mdx)|HTTP and TCP traffic to user-configured endpoints, e.g., internal web applications or the AWS Console.|
|[Teleport Desktop Service](./desktop-access/introduction.mdx)|Remote Desktop Protocol traffic to Windows desktops.|
|[Teleport Kubernetes Service](./kubernetes-access/introduction.mdx)|HTTP traffic to the Kubernetes API server.|

> |[Teleport Node Service](./server-access/introduction.mdx)|Remote shell command execution via an SSH server implementation. Machines you can access via this service are called **Teleport Nodes**.|

Reviewer 2: Let's call this SSH service, since that's how it's referred to when configuring teleport.

---
title: Teleport Core Concepts

> description: Learn the the key architectural components that make up Teleport.

Reviewer 2: ```suggestion
description: Learn the the key components that make up Teleport.
```

---

Here are the core concepts that describe a Teleport deployment. You will see
these terms often when setting up and managing Teleport, so you should get
familiar with them before following other pages in the documentation.

## Teleport cluster

The key concept of Teleport's architecture is the **cluster**. A Teleport
cluster consists of the **Teleport Auth Service** and **Teleport Proxy
Service**, plus the **Teleport Services** that manage traffic to resources
within your infrastructure, such as Kubernetes clusters and Windows desktops.

A minimal Teleport cluster consists of the **Teleport Auth Service** and
**Teleport Proxy Service**. In a demo environment, you can run these two

> services from a single `teleport` binary on a [Linux

Reviewer 2: ```suggestion
services from a single `teleport` process on a [Linux
```

host](./deploy-a-cluster/open-source.mdx).

### Teleport Auth Service

The **Teleport Auth Service** manages **local users** and **configuration
resources** within a **Teleport cluster**. It maintains certificate authorities
that enable users and services to authenticate to your cluster. The Auth Service
issues certificates to clients and maintains an audit log.

The Auth Service is the only component of a cluster that has to be connected to

> a backend, which it uses to store **configuration resources** and the

Reviewer 2: ```suggestion
a backend, which it uses to store **cluster state ** and the
```

and interact with the Auth Service via a gRPC API.

You can run multiple Auth Service instances in a cluster for high availability.

Read our guides to how [authorization](./architecture/authorization.mdx) and
[authentication](./architecture/authentication.mdx) work in Teleport.

### Teleport Proxy Service

The **Teleport Proxy Service** allows for secure access to resources in a your
infrastructure from the public internet without the need for a VPN. 

It establishes reverse tunnels to the **Teleport Auth Service** and **Teleport
Services**, which can run in private networks. This means that, in the Proxy
Service's minimal configuration, you can expose only port `443` to the internet
and run the rest of your infrastructure in private networks.

You can also configure clients to bypass Proxy Service instances and connect to
resources with Teleport-issued certificates directly. 

Read our guide to [how the Teleport Proxy Service
works](./architecture/proxy.mdx).

## Teleport Services

A **Teleport Service** manages access to resources in your infrastructure, such
as Kubernetes clusters, Windows desktops, internal web applications, and
databases.

> A single running `teleport` binary can run one or more **Teleport Services**,

Reviewer 2: ```suggestion
A single running `teleport` process can run one or more **Teleport Services**,
```

Proxies Remote Desktop Protocol traffic to Windows desktops.

Read more about the [Teleport Desktop
Service](./desktop-access/introduction.mdx).

### Teleport Kubernetes Service

Proxies HTTP traffic to the Kubernetes API server.

Read more about the [Teleport Kubernetes
Service](./kubernetes-access/introduction.mdx)

### Teleport SSH Service 

An SSH server implementation that allows users to execute commands on remote
machines while taking advantage of Teleport's built-in access controls,
auditing, and session recording.

Read more about the [Teleport SSH Service](./server-access/introduction.mdx).

### Machine ID

Allows machines and services—called **bot users**—to communicate securely with
resources in your infrastructure by automatically provisioning and renewing
credentials. 

Bot users can connect to resources in your infrastructure without relying
on static credentials (e.g., certificates and private keys) that become more
vulnerable to attacks the longer they remain in use.

Unlike other **Teleport Services**, Machine ID runs via the `tbot` binary,
rather than the `teleport` binary.

Read more in our [Machine ID guide](./machine-id/introduction.mdx).

> ### Agent

Reviewer 2: With the push towards agentless are we sure we want to formally define agent here?

Author: I was looking for a term we could use to describe an instance of a Teleport service. We already use "agent" in the docs, the Teleport source, and conversations with users/prospects, so this seemed appropriate. I know there are misconceptions around the use of agents in Teleport (https://github.com/gravitational/teleport/issues/13351), but I thought we could add clarity to the term here rather than choose a different term.

I also think the term "agent" is better than alternatives. While we could reappropriate "Node" to mean "any service that manages access to a resource", I think the "Node/node" distinction is unwieldy. We could say "Service instance", but it has twice the words/syllables. There's also "relay", as per StrongDM, but that doesn't really describe the SSH Service.

What do you think?

Reviewer 2: I don't feel strongly about this, totally up to you. Agent is fine with me.

## Configuration resources

A **configuration resource** is a document stored on the **Teleport Auth
Service** backend that specifies settings for your **Teleport cluster**.
Examples include **roles**, **local users**, and **authentication connectors**

Read more in our [resource reference](./reference/resources.mdx).

### Role

A **role** is a **configuration resource** that grants **Teleport users**
privileges within a cluster. Teleport's role-based access control (RBAC) is
restrictive by default, and a user needs explicit permissions before they can
access a resource or perform management tasks on a cluster.

Read our [guide to Teleport roles](./access-controls/guides/role-templates.mdx).

### Teleport users

Teleport allows for three kinds of users:

- **Local users** correspond to a `user` **configuration resource** stored on
  the **Auth Service** backend.
- **Single Sign-On (SSO) users** are stored on the backend of your SSO solution,
  e.g., Okta or GitHub. When a user authenticates to Teleport via your SSO
  solution, Teleport issues a certificate for the user and creates a temporary
  **local user** that is valid for the lifetime of the certificate.

> - **Bot users** are automated services that receive credentials via **Machine ID**.

Reviewer 2: Technically there isn't actually a resource that is called bot user. They are the same as local users, they just don't have passwords.

---
title: Teleport Core Concepts
description: Learn the the key architectural components that make up Teleport.
---

> Here are the core concepts describe a Teleport deployment. You will see these

Reviewer 3: ```suggestion
Here are the core concepts that describe a Teleport deployment. You will see these
```

terms often when setting up and managing Teleport, so you should get familiar
with them before following other pages in the documentation.

## Teleport cluster

The key concept of Teleport's architecture is the **cluster**. A Teleport
cluster consists of the **Teleport Auth Service** and **Teleport Proxy Service**,

> plus several **Teleport Services** that manage traffic to resources within your

Reviewer 3: ```suggestion
plus one or more **Teleport Services** that manage traffic to resources within your
```

Author: Suggesting using "the", since our docs sometimes imply that a Teleport cluster consists only of the Auth/Proxy Service

infrastructure, such as Kubernetes clusters and Windows desktops.

A minimal Teleport cluster consists of the **Teleport Auth Service** and
**Teleport Proxy Service**. In a demo environment, you can run these two
services from a single `teleport` binary on a [Linux
host](./deploy-a-cluster/open-source.mdx).

### Teleport Auth Service

The **Teleport Auth Service** manages **local users** and **configuration
resources** within a **Teleport cluster**. It maintains certificate authorities
that enable users and services to authenticate to your cluster. The Auth Service
issues certificates to clients and maintains an audit log.

<Figure
  align="center"
  bordered="false"
  caption="Auth service"
>

> ![Auth service](../img/architecture/auth.png)

Author: I used the images in our original Architectural Overview page here. So we can get this out more quickly, I've suggested removing the image references from this page—maybe we can create a separate issue to fix the images and potentially rework the use of diagrams in the Core Concepts page?

Service](./desktop-access/introduction.mdx).

### Teleport Kubernetes Service

Proxies HTTP traffic to the Kubernetes API server.

Read more about the [Teleport Kubernetes
Service](./kubernetes-access/introduction.mdx)

### Teleport SSH Service 

An SSH server implementation that allows users to execute commands on remote
machines while taking advantage of Teleport's built-in access controls,
auditing, and session recording.

Read more about the [Teleport SSH Service](./server-access/introduction.mdx).

### Machine ID

Allows machines and services—called **bot users**—to communicate securely with
resources in your infrastructure by automatically provisioning and renewing
credentials. 

Bot users can connect to resources in your infrastructure without relying
on static credentials (e.g., certificates and private keys) that become more
vulnerable to attacks the longer they remain in use.

Unlike other **Teleport Services**, Machine ID runs via the `tbot` binary,
rather than the `teleport` binary.

Read more in our [Machine ID guide](./machine-id/introduction.mdx).

### Agent

An instance of a **Teleport Service** is called an **agent**. For all Teleport

> Services besides the **Teleport SSH Service**, an agent  and can enable access

Reviewer 3: ```suggestion
Services besides the **Teleport SSH Service**, an agent can enable access
```
